python/vote-text-bot/bot.py

```diff
@@ -0,0 +1,120 @@
+
+    @type  voter:       str
+    @param voter:       user (display) name
+    @type  vote_count:  int
+    @param vote_count:  current vote count
+    @type  update:      bool
+    @param update:      update existing or create new card?
+    @rtype:             dict
+    @return:            JSON payload to return to Hangouts Chat
+    @see:   developers.google.com/hangouts/chat/concepts/cards
+    """
+
+    PARAMETERS = [{'key': 'count', 'value': str(vote_count)}]
+    return {
+        'actionResponse': {
+            'type': 'UPDATE_MESSAGE' if update else 'NEW_MESSAGE'
+        },
+        'cards': [{
+            'header': {'title': 'Last vote by %s!' % voter},
+            'sections': [{
+                'widgets': [{
+                    'textParagraph': {'text': '%d votes!' % vote_count}
+                }, {
+                    'buttons': [{
+                        'textButton': {
+                            'text': '+1',
+                            'onClick': {
+                                'action': {
+                                    'actionMethodName': 'upvote',
+                                    'parameters': PARAMETERS,
+                                }
+                            }
+                        }
+                    }, {
+                        'textButton': {
+                            'text': '-1',
+                            'onClick': {
+                                'action': {
+                                    'actionMethodName': 'downvote',
+                                    'parameters': PARAMETERS,
+                                }
+                            }
+                        }
+                    }, {
+                        'textButton': {
+                            'text': 'NEW',
+                            'onClick': {
+                                'action': {
+                                    'actionMethodName': 'newvote',
+                                }
+                            }
+                        }
+                    }]
+                }]
+            }]
+        }]
+    }
+
+class MainPage(webapp2.RequestHandler):
+    def post(self):
+        """Handle requests from Hangouts Chat
+
+        * Create new vote for ADDED_TO_SPACE and MESSAGE events
+        * Update existing card for 'upvote' or 'downvote' clicks
+        * Create new vote for 'newvote' clicks
+
+        @type  self:      RequestHandler()
+        @param self:      request/response object
+        """
+        event = json.loads(self.request.body)
+        user = event['user']['displayName']
+        # Update vote when card button pressed
+        if event['type'] == 'CARD_CLICKED':
+            method = event['action']['actionMethodName']
+            # Create new vote when "NEW" button clicked
+            if method == 'newvote':
+                body = createMessage(user)
+            # Update card in-place when "+1" or "-1" button clicked
```

Comment:
I'd put the comment closer to the logic.
```py
else:
    # Update card in-place when "+1" or "-1" button clicked
    delta = 1 ...
```

Reply:
DONE